test/test_log.py

```diff
@@ -156,6 +157,17 @@ def test_long_string_logging(self, backend: FakeModelBackend, logger: logging.Lo
         assert len(calls[1][1]) == MAX_LOG_LINE_LEN
         assert len(calls[2][1]) == 9
 
+    @pytest.mark.filterwarnings("error")
+    def test_warning(self, backend: FakeModelBackend):
+        ops.log.setup_root_logging(backend)
+        with pytest.raises(UserWarning):
+            warnings.warn('this is a test warning')
+            calls = backend.calls()
+            assert len(calls) == 1
+            assert calls[0][0] == 'WARNING'
+            # Match any lineno, in case this file changes in the future.
+            assert re.search(r'test_log\.py:\d+: UserWarning: warning', calls[0][1])
```

Comment:
Shouldn't this end with `UserWarning: this is a test warning`? It seems odd that it would pass as-is, unless I'm missing something?

Reply:
Ha! this line is never executed, because the whole thing is wrapped into `with pytest.raises()` and the code raises early.

`backend.calls()` is actually empty.

Reply:
I dug into this a bit more, and this test just doesn't work. Because of the decorator, `warnings.warn('this is a test warning')` raises an exception, which means that none of the lines after it in the with block ever run. The test passes because the exception is a `UserWarning`, but none of the asserts are ever executed.

For example:

```diff
diff --git a/test/test_log.py b/test/test_log.py
index 2961155..93cb3cf 100644
--- a/test/test_log.py
+++ b/test/test_log.py
@@ -162,6 +162,7 @@ class TestLogging:
         ops.log.setup_root_logging(backend)
         with pytest.raises(UserWarning):
             warnings.warn('this is a test warning')
+            assert False, "this will never be reached"
             calls = backend.calls()
             assert len(calls) == 1
             assert calls[0][0] == 'WARNING'
```

```shell
$ tox -e unit -- -k test_warning
[...]
test/test_log.py::TestLogging::test_warning PASSED                                                                                            [ 25%]
[...]
----------------------------------------------------------------------
TOTAL                       6501   4523   2433     42    26%
  unit: OK (1.36=setup[0.02]+cmd[0.74,0.60] seconds)
  congratulations :) (1.39 seconds)
```

To get the asserts to run, they need to be moved out of the `with`. However, the test will then fail, because it's checking that the warning was captured and sent to the debug log, but that didn't happen, because it was upgraded to an exception, so it didn't go to the logging system.

```diff
diff --git a/test/test_log.py b/test/test_log.py
index 2961155..f737b74 100644
--- a/test/test_log.py
+++ b/test/test_log.py
@@ -162,11 +162,11 @@ class TestLogging:
         ops.log.setup_root_logging(backend)
         with pytest.raises(UserWarning):
             warnings.warn('this is a test warning')
-            calls = backend.calls()
-            assert len(calls) == 1
-            assert calls[0][0] == 'WARNING'
-            # Match any lineno, in case this file changes in the future.
-            assert re.search(r'test_log\.py:\d+: UserWarning: warning', calls[0][1])
+        calls = backend.calls()
+        assert len(calls) == 1
+        assert calls[0][0] == 'WARNING'
+        # Match any lineno, in case this file changes in the future.
+        assert re.search(r'test_log\.py:\d+: UserWarning: warning', calls[0][1])
 
 
 if __name__ == '__main__':
```

```shell
$ tox -e unit -- -k test_warning
[...]
test/test_log.py::TestLogging::test_warning FAILED                                                                                            [ 25%]
[...]
===================================================================== FAILURES ======================================================================
_____________________________________________________________ TestLogging.test_warning ______________________________________________________________
Traceback (most recent call last):
  File "/home/tameyer/scratch/tiexenoperator/.tox/unit/lib/python3.12/site-packages/_pytest/runner.py", line 341, in from_call
    result: Optional[TResult] = func()
                                ^^^^^^
  File "/home/tameyer/scratch/tiexenoperator/.tox/unit/lib/python3.12/site-packages/_pytest/runner.py", line 262, in <lambda>
    lambda: ihook(item=item, **kwds), when=when, reraise=reraise
            ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tameyer/scratch/tiexenoperator/.tox/unit/lib/python3.12/site-packages/pluggy/_hooks.py", line 513, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tameyer/scratch/tiexenoperator/.tox/unit/lib/python3.12/site-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tameyer/scratch/tiexenoperator/.tox/unit/lib/python3.12/site-packages/pluggy/_callers.py", line 182, in _multicall
    return outcome.get_result()
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/tameyer/scratch/tiexenoperator/.tox/unit/lib/python3.12/site-packages/pluggy/_result.py", line 100, in get_result
    raise exc.with_traceback(exc.__traceback__)
  File "/home/tameyer/scratch/tiexenoperator/.tox/unit/lib/python3.12/site-packages/pluggy/_callers.py", line 103, in _multicall
    res = hook_impl.function(*args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tameyer/scratch/tiexenoperator/.tox/unit/lib/python3.12/site-packages/_pytest/runner.py", line 177, in pytest_runtest_call
    raise e
  File "/home/tameyer/scratch/tiexenoperator/.tox/unit/lib/python3.12/site-packages/_pytest/runner.py", line 169, in pytest_runtest_call
    item.runtest()
  File "/home/tameyer/scratch/tiexenoperator/.tox/unit/lib/python3.12/site-packages/_pytest/python.py", line 1792, in runtest
    self.ihook.pytest_pyfunc_call(pyfuncitem=self)
  File "/home/tameyer/scratch/tiexenoperator/.tox/unit/lib/python3.12/site-packages/pluggy/_hooks.py", line 513, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tameyer/scratch/tiexenoperator/.tox/unit/lib/python3.12/site-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tameyer/scratch/tiexenoperator/.tox/unit/lib/python3.12/site-packages/pluggy/_callers.py", line 139, in _multicall
    raise exception.with_traceback(exception.__traceback__)
  File "/home/tameyer/scratch/tiexenoperator/.tox/unit/lib/python3.12/site-packages/pluggy/_callers.py", line 103, in _multicall
    res = hook_impl.function(*args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tameyer/scratch/tiexenoperator/.tox/unit/lib/python3.12/site-packages/_pytest/python.py", line 194, in pytest_pyfunc_call
    result = testfunction(**testargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tameyer/scratch/tiexenoperator/test/test_log.py", line 166, in test_warning
    assert len(calls) == 1
AssertionError: assert 0 == 1
 +  where 0 = len([])
============================================================== short test summary info ==============================================================
FAILED test/test_log.py::TestLogging::test_warning - assert 0 == 1
=================================================== 1 failed, 3 passed, 1179 deselected in 0.53s ====================================================
unit: exit 1 (0.77 seconds) /home/tameyer/scratch/tiexenoperator> coverage run --source=ops/ -m pytest --ignore=test/smoke -v --tb native -k test_warning pid=999475
  unit: FAIL code 1 (0.79=setup[0.02]+cmd[0.77] seconds)
  evaluation failed :( (0.83 seconds)
```

If we're going to test this (and I do think ideally we would) then we need to do it some other way.

Reply:
> Ha! this line is never executed, because the whole thing is wrapped into `with pytest.raises()` and the code raises early.
> 
> `backend.calls()` is actually empty.

Ah, snap :).

Reply:
This does work:

```python
    def test_warning(self, backend: FakeModelBackend):
        ops.log.setup_root_logging(backend)
        warnings.warn_explicit('this is a test warning', UserWarning, 'test_log.py', 100)
        calls = backend.calls()
        assert len(calls) == 1
        assert calls[0][0] == 'WARNING'
        assert re.search(r'test_log\.py:100: UserWarning: this is a test warning', calls[0][1])
```

This isn't precisely testing what happens in prod code, since charms would call `warnings.warn` rather than `warnings.warn_explicit`. However, it's pretty close ([code](https://github.com/python/cpython/blob/0ae8579b85f9b0cd3f287082ad6e194bdb025d88/Lib/warnings.py#L349) with the `warn` code right above).

I think we could also do it by clearing the filters (?) that pytest sets and then restoring them afterwards, but I think just using `warn_explicit` is probably cleaner.

If we really wanted to be sure of what happens we could do a functional test (like the test_main ones) that start up a charm (without Juju) that generates a warning, and have a dummy debug-log hook tool that records use. However, that seems overkill for a small feature like this.

Reply:
Actually, I think it would be fairly simple to add it to [the existing test_main log test](https://github.com/canonical/operator/blob/cf329e60c16ef4f1c36db51daf1a0ffbc19c7e9f/test/test_main.py#L823), just adding in an action to generate a warning, like the [ones that generate logs](https://github.com/canonical/operator/blob/cf329e60c16ef4f1c36db51daf1a0ffbc19c7e9f/test/charms/test_main/src/charm.py#L272). Since that's a subprocess, pytest wouldn't do anything with the warnings. I haven't tested this (I did test `warn_explicit`) but it seems like it would be fine.

Reply:
> This does work:

Maybe not true! I think got sucked in by the "run one test and it works, run them all and it fails" case. However, when I added the test_main case it did pass when they all ran. So I'm a bit :confounded: on that one. Maybe it's the clean-up? You could dig more if you think it's still going to be unstable.

Anyway, this does pass reliably (one test or all) for me:

```diff
--- a/test/charms/test_main/actions.yaml
+++ b/test/charms/test_main/actions.yaml
@@ -29,3 +29,5 @@ log-info:
   description: log a message at Info level
 log-debug:
   description: log a message at Debug level
+warn:
+  description: generate a deprecation warning
diff --git a/test/charms/test_main/src/charm.py b/test/charms/test_main/src/charm.py
index e987d34..ddd80bc 100755
--- a/test/charms/test_main/src/charm.py
+++ b/test/charms/test_main/src/charm.py
@@ -17,6 +17,7 @@ import logging
 import os
 import sys
 import typing
+import warnings
 
 import ops
 
@@ -63,6 +64,7 @@ class Charm(ops.CharmBase):
             on_log_warning_action=[],
             on_log_info_action=[],
             on_log_debug_action=[],
+            on_warn_action=[],
             on_secret_changed=[],
             on_secret_remove=[],
             on_secret_rotate=[],
@@ -107,6 +109,7 @@ class Charm(ops.CharmBase):
             self.framework.observe(self.on.log_warning_action, self._on_log_warning_action)
             self.framework.observe(self.on.log_info_action, self._on_log_info_action)
             self.framework.observe(self.on.log_debug_action, self._on_log_debug_action)
+            self.framework.observe(self.on.warn_action, self._on_warn_action)
 
         self.framework.observe(self.on.collect_metrics, self._on_collect_metrics)
         self.framework.observe(self.on.custom, self._on_custom)
@@ -284,6 +287,9 @@ class Charm(ops.CharmBase):
     def _on_log_debug_action(self, event: ops.ActionEvent):
         logger.debug('insightful debug')
 
+    def _on_warn_action(self, event: ops.ActionEvent):
+        warnings.warn('feature x is going away', DeprecationWarning)
+
     def _on_get_model_name_action(self, event: ops.ActionEvent):
         self._stored._on_get_model_name_action.append(self.model.name)
 
diff --git a/test/test_log.py b/test/test_log.py
index 2961155..ac63458 100644
--- a/test/test_log.py
+++ b/test/test_log.py
@@ -15,6 +15,7 @@
 import io
 import logging
 import re
+import sys
 import typing
 import unittest
 import warnings
@@ -51,6 +52,8 @@ def logger():
     logger = logging.getLogger()
     yield logger
     logging.getLogger().handlers.clear()
+    sys.excepthook = sys.__excepthook__
+    logging.captureWarnings(False)
 
 
 class TestLogging:
@@ -157,16 +160,13 @@ class TestLogging:
         assert len(calls[1][1]) == MAX_LOG_LINE_LEN
         assert len(calls[2][1]) == 9
 
-    @pytest.mark.filterwarnings("error")
     def test_warning(self, backend: FakeModelBackend):
         ops.log.setup_root_logging(backend)
-        with pytest.raises(UserWarning):
-            warnings.warn('this is a test warning')
-            calls = backend.calls()
-            assert len(calls) == 1
-            assert calls[0][0] == 'WARNING'
-            # Match any lineno, in case this file changes in the future.
-            assert re.search(r'test_log\.py:\d+: UserWarning: warning', calls[0][1])
+        warnings.warn_explicit('this is a test warning', UserWarning, 'test_log.py', 100)
+        calls = backend.calls()
+        assert len(calls) == 1
+        assert calls[0][0] == 'WARNING'
+        assert re.search(r'test_log\.py:100: UserWarning: this is a test warning', calls[0][1])
 
 
 if __name__ == '__main__':
diff --git a/test/test_main.py b/test/test_main.py
index 48cfbbc..582e5e0 100644
--- a/test/test_main.py
+++ b/test/test_main.py
@@ -869,6 +869,26 @@ class _TestMain(abc.ABC):
             self._simulate_event(fake_script, event_spec)
             assert calls in fake_script.calls(clear=True)
 
+        # This case is annoying to test, because the message changes depending on which
+        # class is executing the test.
+        event_spec = EventSpec(
+            ops.ActionEvent,
+            'warn_action',
+            env_var='JUJU_ACTION_NAME',
+            set_in_env={'JUJU_ACTION_UUID': '5'},
+        )
+        self._simulate_event(fake_script, event_spec)
+        calls = fake_script.calls(clear=True)
+        for call in calls:
+            # TODO: Probably should check this with a regex to see that there is
+            # something like a filename and a line number.
+            if call[:4] == ['juju-log', '--log-level', 'WARNING', '--'] and call[4].endswith(
+                ': DeprecationWarning: feature x is going away'
+            ):
+                break
+        else:
+            pytest.fail(f'Warning was not sent to debug-log: {calls!r}')
+
     @pytest.mark.usefixtures('setup_charm')
     def test_excepthook(self, fake_script: FakeScript):
         with pytest.raises(subprocess.CalledProcessError):
```